include/cpp-statsd-client/StatsdClient.hpp

```diff
@@ -116,33 +132,52 @@ inline void StatsdClient::send(const std::string& key,
                                const int value,
                                const std::string& type,
                                const float frequency) const noexcept {
-    const auto isFrequencyOne = [](const float frequency) noexcept {
-        constexpr float epsilon{0.0001f};
-        return std::fabs(frequency - 1.0f) < epsilon;
-    };
-
-    // Test if one should send or not, according to the frequency rate
-    if (!isFrequencyOne(frequency)) {
-        if (frequency < static_cast<float>(std::rand()) / RAND_MAX) {
-            return;
-        }
+    constexpr float epsilon{0.0001f};
+    const bool isFrequencyOne = std::fabs(frequency - 1.0f) < epsilon;
+
+    // If you are sampling at a rate less than 1 (ie not sending every metric) and the RNG is above the sampling rate
+    // then we don't need to send this metric this time
+    if (!isFrequencyOne && (frequency < std::uniform_real_distribution<float>(0.f, 1.f)(m_random_engine))) {
+        return;
     }
 
-    // Prepare the buffer, with a sampling rate if specified different from 1.0f
-    char buffer[256];
-    if (isFrequencyOne(frequency)) {
+    // Prepare the buffer and include the sampling rate if it's not 1.f
+    std::string buffer(256, '\0');
+    int string_len = -1;
+    if (isFrequencyOne) {
         // Sampling rate is 1.0f, no need to specify it
-        std::snprintf(buffer, sizeof(buffer), "%s%s:%d|%s", m_prefix.c_str(), key.c_str(), value, type.c_str());
+        string_len = std::snprintf(
+            &buffer.front(), buffer.size(), "%s%s:%d|%s", m_prefix.c_str(), key.c_str(), value, type.c_str());
     } else {
         // Sampling rate is different from 1.0f, hence specify it
-        std::snprintf(
-            buffer, sizeof(buffer), "%s%s:%d|%s|@%.2f", m_prefix.c_str(), key.c_str(), value, type.c_str(), frequency);
+        string_len = std::snprintf(&buffer.front(),
+                                          buffer.size(),
+                                          "%s%s:%d|%s|@%.2f",
+                                          m_prefix.c_str(),
+                                          key.c_str(),
+                                          value,
+                                          type.c_str(),
+                                          frequency);
     }
 
+    // A valid metric must be at least 4 chars in length ie :n|c
+    if (string_len < 4) {
+        //TODO: we dont have access to the error message here
```

Comment:
i feel like this and other scenarios make exceptions feel pretty natural for error handling in this library. we could make a small struct that holds a numeric error number and the string and throw that on error rather than have the caller always check the result. i'm just wondering if maybe you had a specific use-case in mind that maybe `noexcept` were important?

though for this particular error i can remove it completely if i can refactor how the type is selected.. let me work on that

Reply:
makes sense to me!